src/imitation/data/dataset.py

```diff
@@ -0,0 +1,168 @@
+import abc
+import dataclasses
+from typing import Dict, Generic, Mapping, Optional, Tuple, Type, TypeVar
+
+import numpy as np
+
+from imitation.data import types
+
+T = TypeVar("T")
+
+
+class Dataset(abc.ABC, Generic[T]):
+    @abc.abstractmethod
+    def sample(self: "Dataset[T]", n_samples: int) -> T:
+        """Return a batch of data.
+
+        Args:
+            n_samples: A positive integer indicating the number of samples to return.
+        Raises:
+            ValueError: If n_samples is nonpositive.
+        """
+
+    @abc.abstractmethod
+    def size(self: "Dataset[T]") -> Optional[int]:
+        """Returns the number of samples in this dataset, ie the epoch size.
+
+        Returns None if not known or undefined.
```

Comment:
Yep, that was the intention. I'm down to revisit or clarify in the comments

Reply:
This isn't entirely relevant any more because I found out that this isn't True, but for a while I thought that MineRLDatasets wouldn't have easy access to the number of samples in an epoch.

Reply:
I was considering another refactor in BC that would allow training by a number of timesteps in addition to a number of epochs, but I don't think I need this any more so it's on the back burner